src/edu/washington/escience/myria/operator/network/distribute/DistributeFunction.java

```diff
@@ -0,0 +1,82 @@
+package edu.washington.escience.myria.operator.network.distribute;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nonnull;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.google.common.collect.Lists;
+
+import edu.washington.escience.myria.storage.TupleBatch;
+
+/**
+ * A dataset is distributed by two steps: First, using a partition function to generate a partition for each tuple;
+ * Second, mapping each partition to a set of destinations. A destination corresponds to an output channel, which is
+ * determined by a worker ID + an operator ID.
+ */
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+@JsonSubTypes({
+  @Type(value = BroadcastDistributeFunction.class, name = "Broadcast"),
+  @Type(value = HyperCubeDistributeFunction.class, name = "HyperCube"),
+  @Type(value = HashDistributeFunction.class, name = "Hash"),
+  @Type(value = RoundRobinDistributeFunction.class, name = "RoundRobin"),
+  @Type(value = IdentityDistributeFunction.class, name = "Identity")
+})
+public abstract class DistributeFunction implements Serializable {
+
+  /** Required for Java serialization. */
+  private static final long serialVersionUID = 1L;
+
+  /** The partition function. */
+  protected PartitionFunction partitionFunction;
+
+  /** The mapping from partitions to destinations. */
+  protected List<List<Integer>> partitionToDestination;
+
+  /** number of destinations. */
+  protected int numDestinations;
```

Comment:
I think redundancy can only be justified by some cost savings, and I can't imagine that there's any measurable computational cost savings here. And if this is logically a derived quantity (even if you've cached the result), then it doesn't make sense for there to be a setter for it.